### ISSUE TITLE
Improved latexmk support

### ISSUE THREAD

**Issue by hpr1999** [James-Yu/LaTeX-Workshop#2479]
First things first, thanks for the great extension that is currently powering my thesis.
I figure what I'm asking for isn't a minor addition, but I'll humbly make a suggestion:
(also, I'm sorry if there is a way to do this already, I have not found one.)

**Is your feature request related to a problem? Please describe.**
I'm currently using `latexmk -pvc` to compile my documents in order to automatically recompile the document on changes, even with custom dependencies. (For example, I'm using `pandoc` to integrate **markdown** files into my document)
However, the extension seemingly can not  cope with the fact that `latexmk -pvc` does not exit when the document is compiled.
This means that errors aren't highlighted until manually killing the compilation process.
In addition to that, the build analysis just continues to rack up many more phases than a compilation usually has.

**Describe the solution you'd like**

At the basic level, it would be nice if the extension were able to run `latexmk` continuosly and extract error messages and possibly also identify when compilation starts anew instead of simply adding more and more phases in the compilation info.
It would also be nice if the extension updated the progress indicator according to whether latexmk was actually working or just being idle.

Further along, the extension could possibly provide syntax-support for `.latexmkrc` files (I believe it is simply perl-syntax).

**Describe alternatives you've considered**
Another way around my core problem would be to parse dependencies based on the `.latexmkrc` file in which dependencies are defined. That way, the existing auto build functionality could be used with a normal latexmk-invocation.

Of course, in the mean time, I can run `latexmk` externally or simply live without the (very, very nice) error-parsing.

**Additional context**
Example for custom dependencies in a `.latexmkrc` file:

```perl
# Convert Markdown-Files into Latex for inclusion with input or include
add_cus_dep('md', 'tex', 0, 'md2latex');
sub md2latex {
    system("pandoc $_[0].md -f markdown -t latex -o $_[0].tex");
}
```

Overloaded phase-info:

![image](https://user-images.githubusercontent.com/26067274/104757050-deb5db80-575c-11eb-84a6-35999106fcd4.png)

![image](https://user-images.githubusercontent.com/26067274/104757080-e83f4380-575c-11eb-812d-7d958cb9a186.png)

**Comment by jlelong** [comment]
There is no chance that the extension can use a continuously running compilation `latexmk -pvc` and automatically update errors. Yet, I do not understand why you cannot use the auto build feature with the `latexmk` (or `latexmk_rconly`) recipe.

Is your issue that you would like to call LaTeX build from a `.md` file? If so, this is clearly out of the scope of the extension.

**Comment by hpr1999** [comment]
Yes, my use-case was automatically rebuilding on changes in files used to generate dependencies. (markdown to tex, markdown to  pdf, mermaid to images and so on).
I suspected it might be out of scope, but thanks for taking your time to reply anyway, I appreciate it.

**Comment by jlelong** [comment]
Thinking about all this again, I have the feeling it should almost already work with `latex-workshop.latex.autoBuild.run: onFileChange`. Can you give me a complete but short example so that I can make a few tests and understand why it is not working?

**Comment by hpr1999** [comment]
Hey, I've put together a vscode-workspace with an example and wrote up my understanding of this issue in the document.
I tested this configuration with texlive on linux and it requires pandoc to convert markdown to tex.

Please let me know if there's anything else.

[latex-demo.zip](https://github.com/James-Yu/LaTeX-Workshop/files/5823252/latex-demo.zip)

**Comment by jlelong** [comment]
The file `input.md` is not listed in the `main.fls`, hence it is not watched by the extension. So, changes in a `.md` do not trigger a build. `latexmk` provides an other interesting file `main.fdb_latexmk`, which seem to contain an exhausted list of dependencies. I have to at it closely and see if it is relevant for LW to parse this file to discover extra dependencies. 

Yet, the true question is: would it be possible to have the `.md` dependencies properly listed in the `.fls` file by `pandoc`?

**Comment by hpr1999** [comment]
I would be surprised if `pandoc` allowed for a way to do this, since pandoc isn't specifically built for use in `latexmk`, but a general purpose markup converter. In any way, that would seem like the wrong way to me anyway, since it would not help other conversion programs.

However, the conversion process can call arbitrary scripts as defined in the `.latexmkrc` file, so I'll experiment if I could append the converted file to that `fls` file you mentioned as a workaround. I'm not sure if that hack will affect `latexmk` in any other way though, I just don't know enough about it.

Using the `main.fbd_latexmk` sounds like it would be potentially cleaner.

Thanks for checking it out further. :)

**Comment by jlelong** [comment]
The `.fls` files are  generated by the LaTeX compilers when called with the `-recorder` option. They are not specific to `latexmk` unlike the `.fbd_latexmk` file. 

I have just come across the [markdown package](https://ctan.org/pkg/markdown), which handles `.md` to `.tex` conversion on the fly. When using this package, the `.md` files are listed as input files in the `.fls` file and so auto-build works fine when the `.md` files are modified. It may be an interesting alternative.

The difficulty with `pandoc` is that the conversion is not handled by a LaTeX package but by an external command -- be it automatic thanks to the `latexmk` configuration.

**Comment by hpr1999** [comment]
> The difficulty with pandoc is that the conversion is not handled by a LaTeX package but by an external command -- be it automatic thanks to the latexmk configuration.

I agree, but that's also its strength. Thanks for the suggestion, but I'm not only using it to convert `markdown` to `tex`, but also to convert [mermaid](https://mermaid-js.github.io/mermaid/#/) to PDF for graphs, for example.
The possibilities are vast.

**Comment by hpr1999** [comment]
I'm writing this as a separate comment, as I don't know how relevant or interesting it is. 
If it isn't feel free to mark it as offtopic.

Anyway, thanks to you, I have _almost_ managed to get a cumbersome hack working.
In my `.latexmkrc` file, I have added the following:

```perl
sub md2latex {
    system("pandoc $_[0].md -f markdown -t latex -o $_[0].tex");
    system("echo INPUT $_[0].md >> main.fls");
}
```

That way, the extension auto-builds on changing the markdown build.
However, as latexmk when not running in continuous mode doesn't seem to understand that it needs to run the custom dependencies again, I was somewhat stuck.

With `latexmk -c`, I can configure it to remove the generated files as soon as they aren't needed anymore, but I haven't found a way to do that with the extension yet.
(I can add it to the recipe, but that stops LW from reading the `.fls` file first.)

The only thing to make my hack working with this now is to clean them by glob-pattern, which surprisingly works, but might be restrictive to putting files in certain directories, so you don't accidentally delete wrong files.

In this workspace, I can edit the `input.md` and it triggers an auto-update consistently, but I'm afraid it might break on compilations with more than one step (which would be almost all real cases, I think)
[latex-demo.zip](https://github.com/James-Yu/LaTeX-Workshop/files/5825166/latex-demo.zip)

**Comment by tamuratak** [comment]
Stop using the same `.latexmkrc` to compile `main.tex` to `main.pdf` and to convert `input.md` into `input.tex`. Using different `.latexmkrc` solves the issue.

LaTeX Workshop can detect changes on `input.tex`  when you open `main.tex` and compile it. When you convert `input.md` to `input.tex`, the extension can detect the changes on `input.tex` and it compile `main.tex`.

**Comment by hpr1999** [comment]
> When you convert input.md to input.tex with any tools, the extension can detect the changes on input.tex, and it compiles main.tex.

I'm not sure I can follow you. The whole point of these custom dependencies is to _not_ convert them manually.

Anyway, thanks @jlelong for considering it but I suppose this is considered out of scope then.

**Comment by jlelong** [comment]
I have investigated this a little bit further. Using

```perl
sub md2latex {
    system("pandoc $_[0].md -f markdown -t latex -o $_[0].tex");
    system("echo INPUT $_[0].md >> main.fls");
}
```

cannot work because the file `main.fls` is completely overwritten by the LaTeX compiler so there is no way to hack the file `.latexmkrc` to add output to `main.fls`.

An easy way to fix your problem is to pretend to input `input/input.md` so that when the extension parses your project it considers the `.md` file as a dependency. For instance, you can do 

```tex
\iffalse\input{input/input.md}\fi
```

right before

```tex
\input{input/input.tex}
```

**Comment by jlelong** [comment]
Alternatively, you can prefer to add the `.md` dependencies to `main.fls`, by adding

```tex
\newread\tmpread
\newcommand\recorder[1]{%
    \openin\tmpread=#1
    \closein\tmpread
}
\def\mdinput#1{\recorder{#1.md}\input{#1.tex}}
```

and then input `.md` files using `\mdinput{input/input}` without the `.md` extension.

The solution comes from https://tex.stackexchange.com/questions/245336/how-can-one-manually-add-a-line-to-the-pdflatex-recorder-fls-file.

**Comment by hpr1999** [comment]
Thank you so much for investigating further, I'll definitely try these options.

> cannot work because the file main.fls is completely overwritten by the LaTeX compiler so there is no way to hack the file .latexmkrc to add output to main.fls.

I thought the same, but my short tests were definitely successful. But it's probably not consistent.

> ```latex 
> \iffalse\input{input/input.md}\fi
> ```
Am I right to assume that I wouldn't be able to wrap this in a command, as LW only looks for this directly?
Anyway, it definitely works as is with my tests, thanks.

The second one works almost perfectly.
I've defined a command as follows to make it more flexible and it's pretty good.

```latex
 \newread\tmpread
    \newcommand\recorder[1]{%
        \openin\tmpread=#1
        \closein\tmpread
    }
   \newcommand{\inputD}[3]{\recorder{#2.#1}\input{#2.#3}}
    \inputD{md}{input/input}{tex}
```

(Anyway, the only nitpick with the hack is that I can't use my file-autocomplete now, but I don't know if it would even make sense to be able to configure commands for that, let alone how that works)

Thanks again for the research, it's really more than I could've expected.

**Comment by jlelong** [comment]
You are right, you cannot wrap `\iffalse\input{input/input.md}\fi` inside a macro.

You can redefine `\input` so that you can keep using it in a very standard way. With the following code, you just have to use `\input{myfile.md}` and it will be declared as a dependency in `main.fls` and the converted `.tex` file will actually be included.

```tex
\newread\tmpread
\newcommand\recorder[1]{%
    \openin\tmpread=#1
    \closein\tmpread
}

\makeatletter
\let\input@orig\input
\def\input#1{%
    \def\@mdsuffix{md}
    \filename@parse{#1}
    \ifx\filename@ext\@mdsuffix%
        \recorder{#1}%
        \input@orig{\filename@area\filename@base.tex}
    \else%
        \input@orig{#1}%
    \fi%
}
\makeatother
```

**Comment by hpr1999** [comment]
I must admit I'm a novice in "Tex-Programming".
I had actually thought about redefining `input`, but I'd need to be able to pass some optional arguments since `md->tex` is not the only conversion I use. (Examples: `md -> pdf` with includegraphcis, **mmd** --> `pdf`)

For now I have stuck with manually marking the dependencies with this part
```latex
\newread\tmpread
\newcommand\recorder[1]{%
    \openin\tmpread=#1
    \closein\tmpread
}
```
With that, I have just moved my thesis to use the extension instead of `latexmk -pvc` so it definitely stood the further test.
Your redefined `input` looks intriguing, but I'd have to fiddle with it and also make a includegraphics version.

Considering how much time I've already spent on tooling instead of actually writing my thesis, I'll postpone that fiddling.

Thanks for your efforts, now I can see all my errors within VSCode, in all their glory.
